src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -82,9 +84,159 @@ InvokeHostFunctionOpFrame::doApply(AbstractLedgerTxn& ltx)
     throw std::runtime_error("InvokeHostFunctionOpFrame::doApply needs Config");
 }
 
+struct HostFunctionMetrics
+{
+    medida::MetricsRegistry& mMetrics;
+
+    bool mPreflight{false};
+    char const* mFnType{nullptr};
+
+    size_t mReadEntry{0};
+    size_t mWriteEntry{0};
+
+    size_t mReadKeyByte{0};
+    size_t mWriteKeyByte{0};
+
+    size_t mReadDataByte{0};
+    size_t mWriteDataByte{0};
+
+    size_t mReadCodeByte{0};
+    size_t mWriteCodeByte{0};
+
+    size_t mEmitEvent{0};
+    size_t mEmitEventByte{0};
+
+    size_t mCpuInsn{0};
+    size_t mMemByte{0};
+
+    bool mSuccess{false};
+
+    HostFunctionMetrics(medida::MetricsRegistry& metrics, HostFunction hf,
+                        bool isPreflight)
+        : mMetrics(metrics), mPreflight(isPreflight)
+    {
+        if (hf == HOST_FN_INVOKE_CONTRACT)
+        {
+            if (isPreflight)
+            {
+                mFnType = "invoke-preflight";
+            }
+            else
+            {
+                mFnType = "invoke-contract";
+            }
+        }
+        else
+        {
+            if (isPreflight)
+            {
+                mFnType = "create-preflight";
+            }
+            else
+            {
+                mFnType = "create-contract";
+            }
+        }
+    }
+
+    bool
+    isCodeKey(LedgerKey const& lk)
+    {
+        return lk.type() == CONTRACT_DATA &&
+               lk.contractData().key.type() == SCValType::SCV_STATIC &&
+               lk.contractData().key.ic() == SCS_LEDGER_KEY_CONTRACT_CODE;
+    }
+
+    void
+    noteReadEntry(LedgerKey const& lk, size_t n)
+    {
+        mReadEntry++;
+        mReadKeyByte += xdr::xdr_size(lk);
+        if (isCodeKey(lk))
+        {
+            mReadCodeByte += n;
+        }
+        else
+        {
+            mReadDataByte += n;
+        }
+    }
+
+    void
+    noteWriteEntry(LedgerKey const& lk, size_t n)
+    {
+        mWriteEntry++;
+        mWriteKeyByte += xdr::xdr_size(lk);
+        if (isCodeKey(lk))
+        {
+            mReadCodeByte += n;
+        }
+        else
+        {
+            mReadDataByte += n;
+        }
+    }
+
+    ~HostFunctionMetrics()
+    {
+        mMetrics.NewMeter({"host-fn", mFnType, "read-entry"}, "entry")
```

Comment:
If I remember correctly - doing `NewMeter` often like this is very slow and should be avoided (because medida). We probably do not care right now, but will probably show up when we profile later on

Reply:
Yeah, I made this local counter-accumulating object in part to push it back to "one call per metric" rather than "one per entry" but caching the metrics object references higher up be .. a bit more plumbing. I guess I could make a different (non `medida::MetricsRegistry`) object like `OpMetrics` that cache such references, and is passed in and out of each op method, similar to the way `OverlayMetrics` caches them. But for now there's only one op per tx, so this would have to be living up above the tx level -- spanning txs -- to make a difference, and it seemed likely to me not to be worth the effort (these ops are already very very costly, it's probably lost in the noise).